flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/writer/RocksDBSSTIngestWriter.java

```diff
@@ -0,0 +1,238 @@
+		throws IOException {
+		this.db = rocksDB;
+		this.maxSstSize = maxSstSize;
+		this.envOptions = envOptions;
+		this.options = options;
+
+		// Set up a temporary directory for writing generated SST files.
+		// Either, use the provided temporary directory (such as during tests), or create a new one.
+		if (tempDir != null) {
+			this.ingestionTempDir = tempDir;
+		} else {
+			// @lgo: clean the temporary folder up. This may not actually be bad considering
+			// ingestion will move the temporary sst files out.
+			this.ingestionTempDir = Files.createTempDirectory("rocksdb-sst-writer-temp-").toFile();
+		}
+		maxSstSize = 0;
+	}
+
+	public void put(
+		@Nonnull ColumnFamilyHandle columnFamilyHandle, @Nonnull byte[] key, @Nonnull byte[] value)
+		throws RocksDBException, IOException {
+		// Get the sst writer for the column family.
+		RocksDBSSTWriter writer = ensureSSTableWriter(columnFamilyHandle);
+		// Insert the k/v.
+		writer.put(key, value);
+		// Flush the sst and ingest it, if it needs to be flushed.
+		flushIfNeeded(columnFamilyHandle, writer);
+	}
+
+	private RocksDBSSTWriter ensureSSTableWriter(@Nonnull ColumnFamilyHandle columnFamilyHandle)
+		throws RocksDBException, IOException {
+		// Return an existing writer if there is one, Otherwise prepare a new one.
+		if (columnFamilyWriters.containsKey(columnFamilyHandle.getID())) {
+			return columnFamilyWriters.get(columnFamilyHandle.getID());
+		}
+
+		// Create a new sst file in the temporary folder.
+		final String sstFileName = "ingest_" + new AbstractID() + ".sst";
+		File sstFile = new File(ingestionTempDir, sstFileName);
+
+		// Initialize the sst writer.
+		RocksDBSSTWriter writer =
+			new RocksDBSSTWriter(envOptions, options, columnFamilyHandle, sstFile);
+
+		// Store the new writer for the column family.
+		columnFamilyWriters.put(columnFamilyHandle.getID(), writer);
+
+		return writer;
+	}
+
+	/**
+	 * Flushes the data for a particular {@link RocksDBSSTWriter} into the {@link RocksDB} instance.
+	 * After flushing, the {@link ColumnFamilyHandle} will not have an active {@link
+	 * RocksDBSSTWriter}, and it will need to be initialized by {@link
+	 * #ensureSSTableWriter(ColumnFamilyHandle)}.
+	 *
+	 * @throws RocksDBException
+	 */
+	private void flushAndCloseWriter(@Nonnull RocksDBSSTWriter writer) throws RocksDBException {
+		// Finish the sst writer.
+		writer.finish();
+
+		// Instruct RocksDB to ingest the sst files. Because all of the ingested files are
+		// for different column families and the JNI SstFileWriter does not allow setting the
+		// ColumnFamilyHeader when writing the file, need to call ingestExternalFile for one
+		// sst file at a time.
+		//
+		// Because the IngestExternalFileOptions specifies to move the sst file, we do not need
+		// to clean up the written file.
+		List<String> files = Collections.singletonList(writer.getFile().getAbsolutePath());
+		db.ingestExternalFile(writer.getColumnFamilyHandle(), files, INGEST_EXTERNAL_FILE_OPTIONS);
```

Comment:
TODO: I believe this needs `ColumnFamilyOptions` to be passed through. If not here, somewhere else on this write-path.